### PR TITLE
OG-207 commit/reveal penalization

### PR DESCRIPTION
**Pull request by drortirosh**
penalizer doesn't have to be a relayer:
any sender can call `commit()` and then penalize.
(relayers can still penalize without commit)

### CODE REVIEW

File: contracts/Penalizer.sol
Comment (drortirosh) [inline contracts/Penalizer.sol:14]: first time to upgrade version: added functionality, and  fully backward-compatible..

File: contracts/Penalizer.sol
Comment (forshtat) [inline contracts/Penalizer.sol:37]: I believe it would be beneficial if the hash was same for the same penalizeable event regardless of the sender. This way, this method could maybe emit the 'committed' event and then whoever wants to penalize can check is someone has already commited the same fraud proof, saving ether to potentially huge number of people.

Later, we could have some first to commit - first to get paid logic, but this is not necessary for now I think.

File: contracts/Penalizer.sol
Comment (forshtat) [inline contracts/Penalizer.sol:29]: Aslo, please keep the code style at least somewhat consistent. No spaces inside braces, spaces after commas etc.

File: contracts/Penalizer.sol
Comment (forshtat) [inline contracts/Penalizer.sol:44]: There is no requirement to have a delay in this method. This means that the frontrunner can totally put bot commit and reveal in the same block and still outrun the actual 'committer'.

**Review by forshtat — CHANGES_REQUESTED** [review]

File: contracts/Penalizer.sol
Comment (drortirosh) [inline contracts/Penalizer.sol:39]: fix: give "idx" as parameter to commit, and no need for revealSalt.
(that is, the hash of (msgDataHash, msgSender) is done off-chain for the commit, but on-chain for the reveal

File: contracts/Penalizer.sol
Comment (forshtat) [inline contracts/Penalizer.sol:38]: No, don't reveal salt :)

Salt should be revealed in 'reveal' stage.

'hash' should include secret only known to the signer.

File: contracts/Penalizer.sol
Comment (forshtat) [inline contracts/Penalizer.sol:56]: Suggest original path to be removed. 
Registered relays can be added just to front-run penalization, decreasing motivation to monitor chain.

**Review by forshtat — APPROVED** [review]
Includes a lot of my own changes - need a second review.

**Pull request merged**
